pypiserver/backend.py

```diff
@@ -0,0 +1,224 @@
+        raise NotImplementedError
+
+    def add_package(self, filename: str, fh: t.BinaryIO):
+        """Add a package to the Backend. `filename` is the package's filename
+        (without any directory parts). It is just a name, there is no file by
+        that name (yet). `fh` is an open file object that can be used to read
+        the file's content. To convert the package into an actual file on disk,
+        run `as_file(filename, fh)`.
+        """
+        raise NotImplementedError
+
+    def remove_package(self, pkg: PkgFile):
+        """Remove a package from the Backend"""
+        raise NotImplementedError
+
+    def exists(self, filename) -> bool:
+        """Does a package by the given name exist?"""
+        raise NotImplementedError
+
+    def digest(self, pkg: PkgFile):
+        if self.hash_algo is None:
+            return None
+        return digest_file(pkg.fn, self.hash_algo)
+
+    def get_projects(self) -> t.Iterable[str]:
+        """Return an iterable of all (unique) projects available in the store
+        in their PEP503 normalized form. When implementing a Backend class,
+        either use this method as is, or override it with a more performant
+        version.
+        """
+        normalized_pkgnames = set()
+        for x in self.get_all_packages():
+            if x.pkgname:
+                normalized_pkgnames.add(x.pkgname_norm)
+        return normalized_pkgnames
+
+    def find_project_packages(self, project: str) -> t.Iterable[PkgFile]:
+        """Find all packages from a given project. The project may be given
+        as either the normalized or canonical name. When implementing a
+        Backend class, either use this method as is, or override it with a
+        more performant version.
+        """
+        return (
+            x
+            for x in self.get_all_packages()
+            if normalize_pkgname(project) == x.pkgname_norm
+        )
+
+    def find_version(self, name, version) -> t.Iterable[PkgFile]:
+        """Return all packages that match PkgFile.pkgname == name and
+        PkgFile.version == version` When implementing a Backend class,
+        either use this method as is, or override it with a more performant
+        version.
+        """
+        return filter(
+            lambda pkg: pkg.pkgname == name and pkg.version == version,
+            self.get_all_packages(),
+        )
+
+
+class SimpleFileBackend(Backend):
+    def __init__(self, config: Configuration, roots: t.List[PathLike]):
+        super().__init__(config)
+        self.roots = [Path(root).resolve() for root in roots]
+
+    def get_all_packages(self):
+        return itertools.chain.from_iterable(listdir(r) for r in self.roots)
+
+    def add_package(self, filename: str, fh: t.BinaryIO):
+        as_file(fh, self.roots[0].joinpath(filename))
+
+    def remove_package(self, pkg: PkgFile):
+        os.remove(pkg.fn)
+
+    def exists(self, filename):
+        # TODO: Also look in subdirectories?
```

Comment:
I think currently we're recursing into root directories, so we may want to do so here too